### PR TITLE
Add network controller

### PR DESCRIPTION
**Pull request by tuommaki**
Network controller takes care of guest cluster network configuration related responsibilities such as IPAM. This controller is watching for `ClusterNetworkConfig` CRs.

Towards https://github.com/giantswarm/giantswarm/issues/2148

Apiextensions PR must be merged before this: https://github.com/giantswarm/apiextensions/pull/89
Kubernetes apiserver feature flag `CustomResourceSubresources=true` must be also enabled before merging this.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -40,7 +40,7 @@ required = [
   name = "github.com/docker/distribution"
 
 [[constraint]]
-  branch = "master"
+  branch = "add-clusternetwork-crd"
```
Comment (tuommaki) [inline Gopkg.toml]: This shall get reverted back to `master` once `apiextensions` PR gets merged.

File: service/controller/network/v1/key/key.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// ToCustomObject converts value to v1alpha1.ClusterNetworkConfig and returns it or
+// error if type does not match.
+func ToCustomObject(v interface{}) (v1alpha1.ClusterNetworkConfig, error) {
+	customObjectPointer, ok := v.(*v1alpha1.ClusterNetworkConfig)
+	if !ok {
+		return v1alpha1.ClusterNetworkConfig{}, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", &v1alpha1.ClusterNetworkConfig{}, v)
+	}
+
+	if customObjectPointer == nil {
+		return v1alpha1.ClusterNetworkConfig{}, microerror.Maskf(emptyValueError,
```
Comment (xh3b4sd) [inline service/controller/network/v1/key/key.go]: IMO we should not break lines for these things. For me personally this makes readability just worse.

File: service/controller/network/v1/resource_set.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+	resources := []controller.Resource{
+		ipamResource,
+	}
+
+	// Wrap resources with retry and metrics.
+	{
+		retryWrapConfig := retryresource.WrapConfig{}
+
+		retryWrapConfig.BackOffFactory = func() backoff.BackOff { return backoff.WithMaxTries(backoff.NewExponentialBackOff(), ResourceRetries) }
```
Comment (xh3b4sd) [inline service/controller/network/v1/resource_set.go]: Please do this instead. 

```
	{
		c := retryresource.WrapConfig{
			Logger: config.Logger,
		}

		resources, err = retryresource.Wrap(resources, c)
		if err != nil {
			return nil, microerror.Mask(err)
		}
	}

	{
		c := metricsresource.WrapConfig{
			Name: config.ProjectName,
		}

		resources, err = metricsresource.Wrap(resources, c)
		if err != nil {
			return nil, microerror.Mask(err)
		}
	}
```

File: service/controller/network/v1/resource_set.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+		metricsWrapConfig.Name = config.ProjectName
+
+		resources, err = metricsresource.Wrap(resources, metricsWrapConfig)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	initCtxFunc := func(ctx context.Context, obj interface{}) (context.Context, error) {
```
Comment (xh3b4sd) [inline service/controller/network/v1/resource_set.go]: AFAIK the context init function is optional. You can just drop it when you do not need it.

**Comment by tuommaki** [comment]
After discussion on IPAM design, this is not needed anymore. Therefore closing this.

**Pull request closed without merging**